File: backend/src/main/java/com/fullteaching/backend/model/Comment.java
package com.fullteaching.backend.model;

import java.util.List;
import java.util.ArrayList;

import javax.persistence.CascadeType;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import lombok.*;

@Entity
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class Comment {

	@Id
	@GeneratedValue(strategy = GenerationType.AUTO)
	private long id;
	
	private String videourl;
	private boolean audioonly;
	private String message;
	
	private long date;
	
	@OneToMany(mappedBy="commentParent", cascade=CascadeType.ALL)
	@JsonManagedReference
	private List<Comment> replies = new ArrayList<>();
	
	@ManyToOne
	@JsonBackReference
	private Comment commentParent;
	
	@ManyToOne
	private User user;

	public Comment(String message, long date, User user) {
		this.message = message;
		this.date = date;
		this.user = user;
		this.commentParent = null;
	}
	
	public Comment(String message, long date, User user, Comment commentParent) {
		this.message = message;
		this.date = date;
		this.user = user;
		this.commentParent = commentParent;
	}


}

File: frontend/src/app/classes/comment.ts
import { User } from './user';

export class Comment {

  public id?: number;
  public message: string;
  public audioUrl: string;
  public date: Date;
  public replies: Comment[];
  public commentParent: Comment;
  public user: User;

  constructor(message: string, audioUrl: string, commentParent: Comment) {
    this.message = message;
    this.audioUrl = audioUrl;
    this.replies = [];
    this.commentParent = commentParent;
    this.user = null; //Backend will take care of it
  }

}
